bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/XmlSequence.java

```diff
@@ -398,6 +398,23 @@ public XmlValue slice(long startIndex, long endIndex) {
         return new XmlSequence(elementsSeq);
     }
 
+    /**
+     * Search in each element children of the root level xml sequence matching the xml name patterns. When a match is
+     * found, include it in returned sequence and continue the search within the matched element.
+     *
+     * @param qnames Qualified names of the descendants to filter
+     * @return All the descendants that matches the given qualified names, as a sequence, in order of match
+     */
+    public XmlValue rootSequenceDescendants(List<String> qnames) {
```

Comment:
Why we introduce a separate API for this? IMO, this logic should be incorporated within the `descendants` API. When retrieving descendants from a sequence, the elements that are restricted in this context should also be restricted.

Reply:
In the existing implementation the descendants of an xml sequence obtained using the [descendants function](https://github.com/ballerina-platform/ballerina-lang/blob/master/bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/values/XmlSequence.java#L405-L419) in the `XMLSequence` class. 
Consider the case where an element in a sequence contains xml sequence as its children. 
Eg: `<a><b/><c/></a><d/>` (`<a/>` has an xml sequence as it's children).

The current implementation of the `descendants` function is done such that this function is called recursively. The name matching logic of the function atm is done such that it doesn't consider whether the elements in the sequence are top level or nested ones.  Thus when this function is called recursively we need to find a way to identify whether it is called from a top level element or a nested element. 

Since there are no flags in the element to identify whether it is a top level element we need a separate function to identify when a top level element is involved and to ignore the top level elements in such cases. Thus a separate function is needed here. More information is in the [description](https://github.com/ballerina-platform/ballerina-lang/pull/42539#issue-2242600257) of the PR.